### PR TITLE
STOR-1521: Add e2e test cases cover STOR-1443

### PR DESCRIPTION
**Pull request by Phaow**
https://issues.redhat.com/browse/STOR-1443

To run this test manually against an existing cluster:

make && ./openshift-tests run all --dry-run | grep 'Feature:Cluster-CSI-Snapshot-Controller-Operator' | ./openshift-tests run -f -
/cc @openshift/storage-qe

### CODE REVIEW

**Comment by Phaow** [comment]
Helo @jsafrane @mpatlasov, could you please help take a look when you get a chance? Thanks in advance!

File: test/extended/storage/workloads_helpers.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
+package storage
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/workloads_helpers.go:1]: Thanks for the suggestions! Agree, let me update it, some other unused functions I plan to use next several PRs, a better commit should be when I use them then add them.

[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/workloads_helpers.go]: Addessed.

**Comment by Phaow** [comment]
Hi @jsafrane , could you please help take a look again when you get a chance? Thanks! The failed jobs seems not related the added new codes.

File: test/extended/util/deployment.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+// Deployment workload definition
+type Deployment struct {
+	Name             string
+	Namespace        string
+	Replicas         string
+	AppLabel         string
+	MaxWaitReadyTime time.Duration
+}
```
Comment (jsafrane) [inline test/extended/util/deployment.go:25]: IMO we should follow pattern in the other files in `util` directory and expose just functions, not structs. And only functions that are really needed for existing tests, like
`WaitForDeploymentReady` and `GetDeploymentResourceVersion` (or template annotations or generation changed)

See `deploymentconfigs.go` for example.

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		csiSnapshotWebhook                                      exutil.Deployment
+	)
+
+	g.BeforeEach(func() {
+		// Skip if CSISnapshot CO is not enabled
+		if CSISnapshotEnabled, _ := exutil.IsCapabilityEnabled(oc, configv1.ClusterVersionCapabilityCSISnapshot); !CSISnapshotEnabled {
+			g.Skip("Skip for CSISnapshot capability is not enabled on the test cluster!")
+		}
+		clusterCSISnapshotOperatorNs = "openshift-cluster-storage-operator"
+		snapshotWebhookSecretName = "csi-snapshot-webhook-secret"
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:28]: Good point, addressed!

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		defer csiSnapshotWebhook.WaitForReady(oc.AsAdmin())
+		o.Expect(oc.AsAdmin().Run("annotate").Args("-n", clusterCSISnapshotOperatorNs, "secret", snapshotWebhookSecretName,
+			"service.alpha.openshift.io/expiry-", "service.beta.openshift.io/expiry-").Execute()).NotTo(o.HaveOccurred())
+
+		g.By("# Check the webhook Pods should be restarted")
+		o.Eventually(func() bool {
+			csiSnapshotWebhookResourceVersionCurrent := csiSnapshotWebhook.GetSpecifiedJSONPathValue(oc, "{.metadata.resourceVersion}")
+			return strings.EqualFold(csiSnapshotWebhookResourceVersionOri, csiSnapshotWebhookResourceVersionCurrent)
+		}, defaultMaxWaitingTime, defaultMaxWaitingTime/defaultIterationTimes).Should(o.BeFalse(), "The csiSnapshotWebhook deployment resourceVersion doesn't update")
+		csiSnapshotWebhook.WaitForReady(oc.AsAdmin())
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:51]: Agree, removed it.

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	})
+
+	g.AfterEach(func() {
+		WaitForCSOHealthy(oc)
+	})
+
+	g.It("should restart webhook Pods if csi-snapshot-webhook-secret expiry annotation changed", func() {
+
+		g.By("# Get the csiSnapshotWebhook resourceVersion")
+		csiSnapshotWebhookResourceVersionOri := csiSnapshotWebhook.GetSpecifiedJSONPathValue(oc, "{.metadata.resourceVersion}")
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:39]: Good to know, thanks for the details! It seems for the annotations path should be `metadata. annotations `, right? I change to use the `metadata. annotations`.

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+		g.By("# Modify the csi-snapshot-webhook-secret expiry annotation")
+		defer csiSnapshotWebhook.WaitForReady(oc.AsAdmin())
+		o.Expect(oc.AsAdmin().Run("annotate").Args("-n", clusterCSISnapshotOperatorNs, "secret", snapshotWebhookSecretName,
+			"service.alpha.openshift.io/expiry-", "service.beta.openshift.io/expiry-").Execute()).NotTo(o.HaveOccurred())
+
+		g.By("# Check the webhook Pods should be restarted")
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:46]: Updated, thanks!

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+		o.Eventually(func() bool {
+			csiSnapshotWebhookResourceVersionCurrent := csiSnapshotWebhook.GetSpecifiedJSONPathValue(oc, "{.metadata.resourceVersion}")
+			return strings.EqualFold(csiSnapshotWebhookResourceVersionOri, csiSnapshotWebhookResourceVersionCurrent)
+		}, defaultMaxWaitingTime, defaultMaxWaitingTime/defaultIterationTimes).Should(o.BeFalse(), "The csiSnapshotWebhook deployment resourceVersion doesn't update")
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:50]: Updated, thanks!

[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:50]: Good point, updated!

File: test/extended/storage/helpers.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	clusteroperatorhelpers "github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
+	exutil "github.com/openshift/origin/test/extended/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+)
+
+// Define test waiting time const
+const (
+	defaultMaxWaitingTime = 300 * time.Second
+	defaultIterationTimes = 30
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/helpers.go:18]: Updated, thanks!

File: test/extended/util/deployment.go (hunk 2 above)
Comment (Phaow) [reply test/extended/util/deployment.go]: Many thanks for the tips and details! Updated.

**Comment by Phaow** [comment]
> I am sorry I forgot about this PR, I remember you pinged me periodically.

@jsafrane No problem, I know you're always so busy. ^^ I update the codes, could you please help take a look again when you get a chance? Thanks!

File: test/extended/util/deployment.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	pollErr := wait.PollUntilContextTimeout(context.Background(), defaultPollingTime, defaultMaxWaitingTime, true, func(context.Context) (isReady bool, err error) {
+		deployment, getErr = oc.AdminKubeClient().AppsV1().Deployments(namespace).Get(context.Background(), deployName, metav1.GetOptions{})
+		if getErr != nil {
+			e2e.Logf("Unable to retrieve deployment/%s\n %v", deployName, getErr)
+		}
+		if deployment.Status.AvailableReplicas == *deployment.Spec.Replicas {
+			e2e.Logf("Deployment/%s is ready", deployName)
+			return true, nil
+		}
+		e2e.Logf("Deployment/%s is still unready, available replicas %s/%s", deployment.Status.AvailableReplicas, *deployment.Spec.Replicas)
```
Comment (jsafrane) [inline test/extended/util/deployment.go:49]: This produces following log:
```
INFO: Deployment/%!s(int32=1) is still unready, available replicas %!s(int32=2)/%!s(MISSING)
```

* deployName should be the second parameter
* use `%d` (or `%v`) for `AvailableReplicas` / `Replicas`

File: test/extended/util/deployment.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	return pollErr
+}
+
+// DumpDeploymentPodsLogs will dump the deployment pods logs for a deployment for debug purposes
+func DumpDeploymentPodsLogs(oc *CLI, deployName, namespace, labelSelector string) {
+	e2e.Logf("Dumping deployment/%s pods logs \n", deployName)
+
+	pods, err := GetDeploymentPods(oc, deployName, namespace, labelSelector)
+	if err != nil {
+		e2e.Logf("Unable to retrieve pods for deployment %q: %v\n", deployName, err)
```
Comment (jsafrane) [inline test/extended/util/deployment.go:70]: nit: `\n` is not necessary, `Logf` already adds a newline

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+			"service.alpha.openshift.io/expiry-", "service.beta.openshift.io/expiry-").Execute()).NotTo(o.HaveOccurred())
+
+		g.By("# Check the webhook Pods should were restarted")
+		o.Eventually(func() bool {
+			csiSnapshotWebhookAnnotationsCurrent := exutil.GetDeploymentTemplateAnnotations(oc, snapshotWebhookDeployName, clusterCSISnapshotOperatorNs)
+			return reflect.DeepEqual(csiSnapshotWebhookAnnotationsOri, csiSnapshotWebhookAnnotationsCurrent)
+		}).WithTimeout(defaultMaxWaitingTime).WithPolling(defaultPollingTime).Should(o.BeFalse(), "The csiSnapshotWebhook was not updated")
+	})
+
+	g.It("should restart webhook Pods if csi-snapshot-webhook-secret deleted", func() {
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:57]: Updated.

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+		g.By("# Modify the csi-snapshot-webhook-secret expiry annotation")
+		defer func() {
+			if exutil.WaitForDeploymentReady(oc, snapshotWebhookDeployName, clusterCSISnapshotOperatorNs) != nil {
+				e2e.Failf("The csiSnapshotWebhook was not recovered ready")
+			}
+		}()
+		o.Expect(oc.AsAdmin().Run("annotate").Args("-n", clusterCSISnapshotOperatorNs, "secret", snapshotWebhookSecretName,
+			"service.alpha.openshift.io/expiry-", "service.beta.openshift.io/expiry-").Execute()).NotTo(o.HaveOccurred())
+
+		g.By("# Check the webhook Pods should were restarted")
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:50]: Updated.

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+		if CSISnapshotEnabled, _ := exutil.IsCapabilityEnabled(oc, configv1.ClusterVersionCapabilityCSISnapshot); !CSISnapshotEnabled {
+			g.Skip("Skip for CSISnapshot capability is not enabled on the test cluster!")
+		}
+	})
+
+	g.AfterEach(func() {
+		WaitForCSOHealthy(oc)
+	})
+
+	g.It("should restart webhook Pods if csi-snapshot-webhook-secret expiry annotation changed", func() {
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:36]: Updated.

File: test/extended/storage/csi_snapshot_controller_operator.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+		g.By("# Check the webhook Pods should were restarted")
+		o.Eventually(func() bool {
+			csiSnapshotWebhookAnnotationsCurrent := exutil.GetDeploymentTemplateAnnotations(oc, snapshotWebhookDeployName, clusterCSISnapshotOperatorNs)
+			return reflect.DeepEqual(csiSnapshotWebhookAnnotationsOri, csiSnapshotWebhookAnnotationsCurrent)
+		}).WithTimeout(defaultMaxWaitingTime).WithPolling(defaultPollingTime).Should(o.BeFalse(), "The csiSnapshotWebhook was not updated")
+	})
+
+	g.It("should restart webhook Pods if csi-snapshot-webhook-secret deleted", func() {
+
+		g.By("# Get the csiSnapshotWebhook resourceVersion")
```
[the remark this answers is not in the document]
Comment (Phaow) [reply test/extended/storage/csi_snapshot_controller_operator.go:59]: Updated.

File: test/extended/util/deployment.go (hunk 9 above)
Comment (Phaow) [reply test/extended/util/deployment.go:49]: Sorry, I missed it, updated, thanks!

File: test/extended/util/deployment.go (hunk 10 above)
Comment (Phaow) [reply test/extended/util/deployment.go:70]: Yeah, `Logf` already adds a newline, I refer to the deploymentconfigs.go use`\n` seems a bit helpful when debugging with an extra blank line. It is not necessary, removed.

**Comment by Phaow** [comment]
Hello @jsafrane, sorry for so many typos, I update the codes, could you please help take a look again when you get a chance? Thanks!

**Comment by jsafrane** [comment]
/assign @soltysh
for approval

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**